File: NMS/v1_20_R1/src/main/java/dev/rosewood/roseskyblock/nms/v1_20_R1/NMSHandlerImpl.java
package dev.rosewood.roseskyblock.nms.v1_20_R1;

import dev.rosewood.roseskyblock.nms.BorderColor;
import dev.rosewood.roseskyblock.nms.NMSHandler;
import net.minecraft.network.protocol.game.ClientboundInitializeBorderPacket;
import net.minecraft.world.level.border.WorldBorder;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_20_R1.CraftWorld;
import org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer;
import org.bukkit.entity.Player;

public class NMSHandlerImpl implements NMSHandler {

    @Override
    public void sendWorldBorder(Player player, BorderColor color, double size, Location center) {
        WorldBorder border = new WorldBorder();
        border.world = ((CraftWorld) center.getWorld()).getHandle();
        border.setCenter(center.getBlockX() + 0.5, center.getBlockZ() + 0.5);
        border.setWarningBlocks(0);
        border.setWarningTime(0);

        switch (color) {
            case OFF -> border.setSize(Integer.MAX_VALUE);
            case BLUE -> border.setSize(size);
            case RED -> border.lerpSizeBetween(size, size - 1.0, 20000000L);
            case GREEN -> border.lerpSizeBetween(size - 1.0, size, 20000000L);
        }

        ((CraftPlayer) player).getHandle().connection.send(new ClientboundInitializeBorderPacket(border));
    }

}

File: NMS/v1_15_R1/src/main/kotlin/dev/rosewood/roseskyblock/nms/v1_15_R1/NMSHandlerImpl.kt
package dev.rosewood.roseskyblock.nms.v1_15_R1

import dev.rosewood.roseskyblock.nms.BorderColor
import dev.rosewood.roseskyblock.nms.NMSHandler
import net.minecraft.server.v1_15_R1.PacketPlayOutWorldBorder
import net.minecraft.server.v1_15_R1.WorldBorder
import org.bukkit.Location
import org.bukkit.craftbukkit.v1_15_R1.CraftWorld
import org.bukkit.craftbukkit.v1_15_R1.entity.CraftPlayer
import org.bukkit.entity.Player

class NMSHandlerImpl : NMSHandler {

    override fun sendWorldBorder(player: Player, color: BorderColor, size: Double, center: Location) {
        val worldBorder = WorldBorder()
        worldBorder.world = (center.world as CraftWorld).handle
        worldBorder.setCenter(center.blockX + 0.5, center.blockZ + 0.5)

        if (color == BorderColor.OFF) {
            worldBorder.size = Double.MAX_VALUE
        } else {
            worldBorder.size = size
        }
        worldBorder.warningDistance = 0
        worldBorder.warningTime = 0

        if (color == BorderColor.RED) {
            worldBorder.transitionSizeBetween(size, size - 1.0, 20000000L)
        } else if (color == BorderColor.GREEN) {
            worldBorder.transitionSizeBetween(size - 0.1, size, 20000000L)
        }

        (player as CraftPlayer).handle.playerConnection.sendPacket(PacketPlayOutWorldBorder(worldBorder, PacketPlayOutWorldBorder.EnumWorldBorderAction.INITIALIZE))
    }

}
